{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "QCDpred.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "#@title Load utility functions\n",
        "import numpy as np\n",
        "\n",
        "def unique(vec):\n",
        "    seen = set()\n",
        "    seen_add = seen.add\n",
        "    return [v for v in vec if not (v in seen or seen_add(v))]\n",
        "\n",
        "def is_aa_one_nat(sequence, additional=\"\"):\n",
        "  additional = additional.upper()\n",
        "  for aa in sequence.upper():\n",
        "    if not (aa in \"ACDEFGHIKLMNPQRSTVWY\" or aa in additional):\n",
        "      return(False)\n",
        "    return(True)\n",
        "\n",
        "def read_fasta(filename, comment_char=\"#;\"):\n",
        "    \"\"\"Flexible FASTA file reader without dependencies\"\"\"\n",
        "    seq_list = []\n",
        "    file_id = filename.split(\"/\")[-1].split(\".\")[0]\n",
        "    reading_fasta = False\n",
        "    with open(filename,\"r\") as file_handle:\n",
        "        for line in file_handle:\n",
        "            words = line.split()\n",
        "            if len(words) == 0:\n",
        "                continue\n",
        "            if words[0][0] in comment_char:\n",
        "                continue\n",
        "            if words[0][0] == \">\":\n",
        "                if reading_fasta:\n",
        "                    seq_list.append((seq_id,seq))\n",
        "                seq_id = words[0][1:]\n",
        "                seq = \"\"\n",
        "                reading_fasta = True\n",
        "            elif reading_fasta:\n",
        "                if len(words) > 1:\n",
        "                    print(\"WARNING: Found FASTA line with more white space separated fields:\\n%s\" % (line))\n",
        "                seq = seq + words[0]\n",
        "            else:\n",
        "                if len(words) == 1:\n",
        "                    seq = words[0]\n",
        "                    if not is_aa_one_nat(seq):\n",
        "                        print(\"ERROR: Non-FASTA single-column file should have a protein sequence in first column:\")\n",
        "                        print(line)\n",
        "                        return(None)\n",
        "                    seq_id = file_id+\"%05d\" % (len(seq_list))\n",
        "                    seq_list.append((seq_id,seq))\n",
        "                elif len(words) > 1:\n",
        "                    seq = words[1]\n",
        "                    if not is_aa_one_nat(seq):\n",
        "                        print(\"ERROR: Non-FASTA multi-column file should have a protein sequence in second column:\")\n",
        "                        print(line)\n",
        "                        return(None)\n",
        "                    seq_list.append((words[0],seq))\n",
        "        if reading_fasta:\n",
        "            seq_list.append((seq_id,seq))\n",
        "    return(seq_list)\n",
        "\n",
        "def tile_sequence(seq, tile_len, tile_offset=1, add_ct_tile=True):\n",
        "  \"\"\"Cut sequence into tiles of length tile_len\"\"\"\n",
        "  nres = len(seq)\n",
        "  tile_begin = 0\n",
        "  tile_end = tile_len\n",
        "  tile_list = []\n",
        "  while(tile_end <= nres):\n",
        "    tile_list.append(seq[tile_begin:tile_end])\n",
        "    tile_begin = tile_begin + tile_offset\n",
        "    tile_end = tile_begin + tile_len\n",
        "  if tile_end !=  nres+1 and add_ct_tile:\n",
        "    tile_list.append(seq[-tile_len:])\n",
        "  return(tile_list)\n",
        "\n",
        "# Trained on WT data with >50 raw counts total in 4 the bins                                                                                                                        \n",
        "# Peptides classified degrons if PSI<2.2 and stable if PSI>2.8                                                                                                                      \n",
        "# Parameter are per amino acid counts in peptides of length 17                                                                                                                      \n",
        "model = {}\n",
        "model[\"intersect\"] = -0.89102423\n",
        "model[\"C\"] =   0.37721431\n",
        "model[\"D\"] =  -0.78986558\n",
        "model[\"E\"] =  -0.65124014\n",
        "model[\"K\"] =  -0.15518666\n",
        "model[\"R\"] =  -0.02030300\n",
        "model[\"H\"] =  -0.02110156\n",
        "model[\"N\"] =  -0.32782161\n",
        "model[\"Q\"] =  -0.17676485\n",
        "model[\"A\"] =   0.10844211\n",
        "model[\"G\"] =  -0.37594135\n",
        "model[\"S\"] =  -0.09627044\n",
        "model[\"T\"] =  -0.08533912\n",
        "model[\"V\"] =   0.43746326\n",
        "model[\"M\"] =   0.31182498\n",
        "model[\"L\"] =   0.53427787\n",
        "model[\"I\"] =   0.61465146\n",
        "model[\"F\"] =   0.52882600\n",
        "model[\"Y\"] =   0.45253658\n",
        "model[\"W\"] =   0.58693535\n",
        "model[\"P\"] =  -0.25880796\n",
        "\n",
        "def calc_degron_prob(seq_list):\n",
        "  p_list = []\n",
        "  for seq in seq_list:\n",
        "    # This model is only valid for peptides of length 17                                                                                                                        \n",
        "    assert len(seq) == 17\n",
        "    lin = model[\"intersect\"] + sum([model[aa] for aa in seq])\n",
        "    responce = 1/(1+np.exp(-lin))\n",
        "    p_list.append(responce)\n",
        "  return(p_list)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "u9oC5Z13RTbn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Z5c__G6cws8x",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "from google.colab import files\n",
        "import re\n",
        "\n",
        "#@title Input protein sequence(s), then hit `Runtime` -> `Run all`\n",
        "#@markdown Input one amino acid sequence\n",
        "\n",
        "name_input = 'APPY' #@param {type:\"string\"}\n",
        "sequence_input = 'CALLQSRLLLSAPRRAAATARY' #@param {type:\"string\"}\n",
        "\n",
        "#@markdown and/or upload a FASTA file (check box, click run and a dialog will appear blow this).\n",
        "\n",
        "upload_fasta_file = False #@param {type:\"boolean\"}\n",
        "\n",
        "seq_list = []\n",
        "\n",
        "# Read text input\n",
        "if sequence_input != \"\":\n",
        "  # remove non-word characters, i.e. whitespace, tab, newline, etc\n",
        "  sequence = re.sub(r\"\\W\", \"\", sequence_input).upper()\n",
        "  if len(sequence) < 17:\n",
        "    print(\"WARNING: Sequence '%s' is shorter than 17 amino acid and ignored\" % \n",
        "          (sequence))\n",
        "  elif not is_aa_one_nat(sequence):\n",
        "    print(\"WARNING: Sequence '%s' is not a canonical amino acid sequence and ignored\" % \n",
        "          (sequence))\n",
        "  else:\n",
        "    name = name_input.strip().split()[0]\n",
        "    if name == \"\":\n",
        "      name = \"seq_input\"\n",
        "    seq_list.append((name, sequence))\n",
        "\n",
        "# Read file input\n",
        "if upload_fasta_file:\n",
        "  uploaded = files.upload()\n",
        "  for filename in uploaded.keys():\n",
        "    # bug: if the same name.fa is uploaded, this will write name (1).fa \n",
        "    # but not update filename variable so the old file will be loaded!\n",
        "    print(\"Read %s\" % (filename))\n",
        "    for (seq_id,seq_aa) in read_fasta(filename):\n",
        "      if len(seq_aa) < 17:\n",
        "        print(\"WARNING: Sequence %s %s from file %s is shorter than 17 amino acids and ignored\" % \n",
        "              (seq_id,seq_aa,filename))\n",
        "      else:\n",
        "        seq_list.append((seq_id,seq_aa.upper()))\n",
        "\n",
        "if len(seq_list) > 0:\n",
        "  print(\"Got %d sequences from input - in total %d residues\" % \n",
        "        (len(seq_list),sum([len(v[1]) for v in seq_list])))\n",
        "  \n",
        "  # check names\n",
        "  seq_ids = [v[0] for v in seq_list]\n",
        "  if len(seq_ids) != len(set(seq_ids)):\n",
        "    print(\"WARNING: Sequence names used to identify tiles are not unique\")\n",
        "  \n",
        "  # Tile sequences and calculate degron probability\n",
        "  result = []\n",
        "  for (seq_id,seq) in seq_list:\n",
        "    tile_seq = tile_sequence(seq, 17, 1)\n",
        "    tile_degron_prob = calc_degron_prob(tile_seq)\n",
        "    # assign degron probability to central amino acid\n",
        "    resi = 9\n",
        "    for i in range(len(tile_seq)):\n",
        "      result.append([seq_id, tile_seq[i], tile_degron_prob[i], tile_seq[i][8], resi])\n",
        "      # print(\"%18s  %17s  %.5f  %1s  %4d\" % \n",
        "      #       (seq_id, tile_seq[i], tile_degron_prob[i], tile_seq[i][8], resi))\n",
        "      resi += 1\n",
        "else:\n",
        "  print(\"No input\")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import matplotlib.pyplot as plt\n",
        "\n",
        "#@title Plot degron probabilities for a protein\n",
        "#@markdown Give the index or name of a protein to plot\n",
        "\n",
        "index_or_name = '0' #@param {type:\"string\"}\n",
        "\n",
        "seq_names = [v[0] for v in result]\n",
        "seq_names_uniq = unique(seq_names)\n",
        "name2plot = None\n",
        "\n",
        "try:\n",
        "  plot_index = int(index_or_name)\n",
        "  assert plot_index <= len(seq_names_uniq)\n",
        "  name2plot = seq_names_uniq[plot_index]\n",
        "except:\n",
        "  name2plot = index_or_name\n",
        "\n",
        "if name2plot in seq_names_uniq:\n",
        "  plot_index = np.where(np.array(seq_names) == name2plot)[0]\n",
        "  result_np = np.array(result)\n",
        "  resi = [int(v[4]) for v in result_np[plot_index] ]\n",
        "  if (len(resi) != len(set(resi))):\n",
        "    print(\"WARNING: More points for same residue index - are sequence names unique?\")\n",
        "  deg_prob = [float(v[2]) for v in result_np[plot_index] ]\n",
        "  plt.plot(resi, deg_prob, label=name2plot)\n",
        "  plt.plot()\n",
        "\n",
        "  plt.xlabel(\"Residue index\")\n",
        "  plt.ylabel(\"Degron probaility\")\n",
        "  # plt.title(\"Line Graph Example\")\n",
        "  plt.legend()\n",
        "  plt.show()\n",
        "else:\n",
        "  print(\"Cannot plot sequence \\'%s\\'\" % (name2plot))\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "BSmUkI9IquOA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download results\n",
        "#@markdown If this fails, try googles chorme browser\n",
        "filename = 'QCDpred.txt' #@param {type:\"string\"}\n",
        "write_header = True #@param {type:\"boolean\"}\n",
        "\n",
        "if len(result) > 0:\n",
        "  with open(filename, 'w') as f:\n",
        "    if write_header:\n",
        "      # white space separated text\n",
        "      f.write(\"%-20s  %17s  %8s  %2s  %4s\\n\" % \n",
        "              (\"name\",\"sequence\",\"deg_prob\",\"aa\",\"resi\"))\n",
        "      \n",
        "      # ecxel style csv\n",
        "      # f.write(\"name;sequence;deg_prob;aa;resi\")\n",
        "    for line in result:\n",
        "      # white space separated text\n",
        "      f.write(\"%-20s  %17s  %8.5f  %2s  %4d\\n\" % tuple(line))\n",
        "\n",
        "      # quoted strings\n",
        "      # f.write(\"%-20s  %17s  %8.5f  %2s  %4d\\n\" % \n",
        "      #        (\"\\\"\"+line[0]+\"\\\"\", \"\\\"\"+line[1]+\"\\\"\", line[2], \"\\\"\"+line[3]+\"\\\"\", line[4]))\n",
        "      \n",
        "      # ecxel style csv\n",
        "      # f.write(\"%s;%s;%.5f;%s;%d\\n\" % tuple(line))\n",
        "  files.download(filename)\n",
        "else:\n",
        "  print(\"No result to download\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "wypZ6NgNSMrK"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
